### ISSUE TITLE
PHP Proxy does not proxy Content-Type 'text/xml' requests

### ISSUE THREAD

**Issue by dasa** [Esri/resource-proxy#366]
I'm using PHP Version 5.5.34.

Request:

```
POST /PHP/proxy.php?http%3A%2F%2Fsuite.opengeo.org%2Fgeoserver%2Fwfs= HTTP/1.1
Pragma: no-cache
Origin: http://quicksilver.esri.com
Accept-Encoding: gzip, deflate
Host: quicksilver.esri.com
Accept-Language: en-US,en;q=0.8
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.79 Safari/537.36
Content-Type: text/xml
Accept: */*
Cache-Control: no-cache
X-Requested-With: XMLHttpRequest
Cookie: __utma=39465783.1702136191.1446065835.1446065835.1449088386.2; WT_FPC=id=277f886dab32e00d4011449088385181:lv=1449088416334:ss=1449088385181; __unam=b1ba6c-15164660495-3c3d8963-2; s_vnum=1495300342950%26vn%3D1; _ga=GA1.2.1702136191.1446065835; _ceg.s=o7hkyq; _ceg.u=o7hkyq; _ceir=1; s_fid=080762D3DDC74C02-0D9A896E9D205BD8; s_nr=1463765860424-New; PHPSESSID=7d72852cead1c32cae4e5284ddf2e6b4
Connection: keep-alive
Referer: http://quicksilver.esri.com/js-api/test-apps/3.17/webmap.html
Content-Length: 440

<?xml version='1.0' encoding='utf-8'?>
<GetFeature
xmlns:xsi='http://www.w3.org/2001/XMLSchema-instance'
xmlns:gml='http://www.opengis.net/gml'
xmlns:ogc='http://www.opengis.net/ogc'
xmlns:wfs='http://www.opengis.net/wfs'
xmlns='http://www.opengis.net/wfs'
xmlns:opengeo='http://opengeo.org'
version='1.1.0' service='WFS' maxFeatures='3000'>
<wfs:Query typeName='opengeo:countries' srsName='EPSG:4326'>
</wfs:Query>
</GetFeature>
```

Response:

```xml
<ows:ExceptionReport xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:ows="http://www.opengis.net/ows/1.1" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" version="2.0.0" xsi:schemaLocation="http://www.opengis.net/ows/1.1 http://suite.opengeo.org/geoserver/schemas/ows/1.1.0/owsAll.xsd">
<ows:Exception exceptionCode="MissingParameterValue" locator="request">
<ows:ExceptionText>Could not determine geoserver request from http request org.geoserver.platform.AdvancedDispatchFilter$AdvancedDispatchHttpRequest@50728dd</ows:ExceptionText>
</ows:Exception>
</ows:ExceptionReport>
```

**Comment by dasa** [comment]
Here's a .har file:
[wfs-php-proxy.har.zip](https://github.com/Esri/resource-proxy/files/297867/wfs-php-proxy.har.zip)

**Comment by dasa** [comment]
It looks like PR #323 was for a similar issue. Hopefully the fix could be more generic than this:
https://github.com/Yall1963/resource-proxy/commit/e0d9a7df0fe96241455ad2cff9434debb5f27e1c#diff-4d91edb8990afeef04888edc8d4be957

**Comment by tgirgin23** [comment]
Any news on getting this fixed? We are trying to display a WFS service in a ESRI JS 3.20 map. We have tried to get this to work on the PHP proxy, without any success. This is however working fine on the .net proxy.

**Comment by bsvensson** [comment]
__Update__
The php proxy is failing because it's using PHP's built-in `$_POST` to determine if it's a POST.  According to the user comment at http://php.net/manual/en/reserved.variables.post.php#114598, this is only true for standard forms with `Content-Type: application/x-www-form-urlencoded`.

$_POST is NOT set for: `Content-Type:text/xml`.

This is why the logic in the setupClassProperties function is failing. And the PHP proxy isn't working for `application/xml` or `text/xml` etc.

**Comment by bsvensson** [comment]
@tgirgin23 - If your WFS is public, could you share it for testing?

**Comment by tgirgin23** [comment]
@bsvensson Thanks for the quick reply. The WFS layer is the one found in the current [JS API sample WFS page](https://developers.arcgis.com/javascript/3/sandbox/sandbox.html?sample=layers_wfs).

**Comment by bsvensson** [comment]
@tgirgin23 - no problem. BTW, that's just a test service with some test data. I assume you'll be using something else later on (unless this is just a test/prototype app). Since different WFS can be set up quite differently, don't forget to test on the actual service.

**Comment by tgirgin23** [comment]
@bsvensson We have tried on multiple different services and we are having the same issue every single time with the PHP proxy

**Comment by paupepui** [comment]
I have the same problem when I use a WFS layer in a JavaScript ESRI API
I get the following error:

`No query information to decode. QUERY_STRING is set, but empty.`

This is the request:

```
Request URL:http://aplicaciones.icv.gva.es/visor_gva/proxy/PHP/proxy.php?http://mapas.valencia.es/wms/opendata_wfs?
Request Method:POST
Status Code:200 OK
Remote Address:172.16.30.66:80
Referrer Policy:no-referrer-when-downgrade
Response Headers
view source
Access-Control-Allow-Headers:*
Access-Control-Allow-Origin:*
Cache-Control:max-age=1
Connection:close
Content-Encoding:gzip
Content-Length:84
Content-Type:text/html; charset=UTF-8
Date:Mon, 16 Oct 2017 08:46:47 GMT
Expires:Thu, 19 Nov 1981 08:52:00 GMT
Expires:Mon, 16 Oct 2017 08:46:50 GMT
Server:Apache/2.2.29 (Unix) mod_ssl/2.2.29 OpenSSL/1.0.1e-fips DAV/2 PHP/5.5.23
Vary:Accept-Encoding
Request Headers
view source
Accept:*/*
Accept-Encoding:gzip, deflate
Accept-Language:es-ES,es;q=0.8,en;q=0.6,ca;q=0.4,fr;q=0.2
Cache-Control:no-cache
Connection:keep-alive
Content-Length:763
Content-Type:text/xml
Cookie:PHPSESSID=j1m2pgbtpr05ljk4phektgggp7; SESS3bf9382ecbe8ca0f95636584e4e9a0b3=k57u9dc3ttt6pn3d0b8avfo9e6
Host:aplicaciones.icv.gva.es
Origin:http://aplicaciones.icv.gva.es
Pragma:no-cache
Referer:http://aplicaciones.icv.gva.es/visor_gva/visor/index.html?modo=intranet
User-Agent:Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.100 Safari/537.36
X-Requested-With:XMLHttpRequest
Query String Parameters
view source
view URL encoded
http://mapas.valencia.es/wms/opendata_wfs?:
Request Payload
<?xml version='1.0' encoding='utf-8'?>
<GetFeature 
 xmlns:xsi='http://www.w3.org/2001/XMLSchema-instance'
 xmlns:gml='http://www.opengis.net/gml'
 xmlns:ogc='http://www.opengis.net/ogc'
 xmlns:wfs='http://www.opengis.net/wfs'
 xmlns='http://www.opengis.net/wfs'
 xmlns:='http://www.opengis.net/wfs'
 version='1.1.0' service='WFS' maxFeatures='1000'>
 <wfs:Query typeName=':TRA_RECARGA_ELECTRICA' srsName='EPSG:3857'>
  <ogc:Filter xmlns:ogc='http://www.opengis.net/ogc'>
   <ogc:BBOX>
    <ogc:PropertyName>:msGeometry</ogc:PropertyName>
    <gml:Box srsName='EPSG:3857'>
      <gml:coordinates>-436357.58482010243,4692951.475488542 346357.58482010243,4951614.379205579</gml:coordinates>
    </gml:Box>
   </ogc:BBOX>
  </ogc:Filter>
 </wfs:Query>
</GetFeature>
```
Thanks!

**Comment by wanderlux** [comment]
Had the same problem with @paupepui . I could find a workaround valid for us. I forked the repository and added the changes there. It probably needs serious revision, but could be used as a patch meanwhile.  Here is a comparison showing what I added:

https://github.com/Esri/resource-proxy/compare/master...wanderlux:master

TL;DR

- Added a new check for `Content-Type:text/xml` in `setupClassProperties` following @bsvensson hint.
- Added a new option so this function would be called, using a `proxyMethod` I named `XML`
- Added a new function  `proxyXml`: copied `proxyPost` and changed the values I needed, passing the raw xml data into the request body.
